src/sn_utils.c

```diff
@@ -265,61 +267,101 @@ static uint16_t reg_lifetime(n2n_sn_t *sss)
-                       time_t now)
-{
-    macstr_t mac_buf;
-    n2n_sock_str_t sockbuf;
-    struct peer_info *scan;
-
-    traceEvent(TRACE_DEBUG, "update_edge for %s [%s]",
-               macaddr_str(mac_buf, edgeMac),
-               sock_to_cstr(sockbuf, sender_sock));
-
-    HASH_FIND_PEER(comm->edges, edgeMac, scan);
-
-    if (NULL == scan)
-    {
-        /* Not known */
-
-        scan = (struct peer_info *)calloc(1,
-                                          sizeof(struct peer_info)); /* deallocated in purge_expired_registrations */
-
-        memcpy(&(scan->mac_addr), edgeMac, sizeof(n2n_mac_t));
-        memcpy(&(scan->sock), sender_sock, sizeof(n2n_sock_t));
-        scan->last_valid_time_stamp = initial_time_stamp ();
+                       time_t now) {
+	macstr_t mac_buf;
+	n2n_sock_str_t sockbuf;
+	struct peer_info *scan;
+
+	traceEvent(TRACE_DEBUG, "update_edge for %s [%s]",
+	           macaddr_str(mac_buf, reg->edgeMac),
+	           sock_to_cstr(sockbuf, sender_sock));
+
+	HASH_FIND_PEER(comm->edges, reg->edgeMac, scan);
+
+	if (NULL == scan) {
+		/* Not known */
+
+		scan = (struct peer_info *) calloc(1,
+		                                   sizeof(struct peer_info)); /* deallocated in purge_expired_registrations */
+
+		memcpy(&(scan->mac_addr), reg->edgeMac, sizeof(n2n_mac_t));
+		scan->dev_addr.net_addr = reg->dev_addr.net_addr;
+		scan->dev_addr.net_bitlen = reg->dev_addr.net_bitlen;
+		memcpy(&(scan->sock), sender_sock, sizeof(n2n_sock_t));
+		scan->last_valid_time_stamp = initial_time_stamp();
+
+		HASH_ADD_PEER(comm->edges, scan);
+
+		traceEvent(TRACE_INFO, "update_edge created   %s ==> %s",
+		           macaddr_str(mac_buf, reg->edgeMac),
+		           sock_to_cstr(sockbuf, sender_sock));
+	} else {
+		/* Known */
+		if (!sock_equal(sender_sock, &(scan->sock))) {
+			memcpy(&(scan->sock), sender_sock, sizeof(n2n_sock_t));
+
+			traceEvent(TRACE_INFO, "update_edge updated   %s ==> %s",
+			           macaddr_str(mac_buf, reg->edgeMac),
+			           sock_to_cstr(sockbuf, sender_sock));
+		} else {
+			traceEvent(TRACE_DEBUG, "update_edge unchanged %s ==> %s",
+			           macaddr_str(mac_buf, reg->edgeMac),
+			           sock_to_cstr(sockbuf, sender_sock));
+		}
+	}
+
+	scan->last_seen = now;
+	return 0;
+}
 
-        HASH_ADD_PEER(comm->edges, scan);
 
-        traceEvent(TRACE_INFO, "update_edge created   %s ==> %s",
-                   macaddr_str(mac_buf, edgeMac),
-                   sock_to_cstr(sockbuf, sender_sock));
-    }
-    else
-    {
-        /* Known */
-        if (!sock_equal(sender_sock, &(scan->sock)))
-        {
-            memcpy(&(scan->sock), sender_sock, sizeof(n2n_sock_t));
+static signed int peer_tap_ip_sort(struct peer_info *a, struct peer_info *b) {
+	uint32_t a_host_id = a->dev_addr.net_addr & (~bitlen2mask(a->dev_addr.net_bitlen));
+	uint32_t b_host_id = b->dev_addr.net_addr & (~bitlen2mask(b->dev_addr.net_bitlen));
+	return ((signed int)a_host_id - (signed int)b_host_id);
+}
```

Comment:
As we will have all different kind of sub-networks, we might need to sort by whole IP address. So, just to be sure, as int is 32 bit: Check if this works the same way with all sub-networks. Especially clarify if those with  MSB reset (e.g. 10.x.x.x) are treated equal to those with MSB set (e.g. 192.168.x.x).